docs/source/components.rst

```diff
@@ -0,0 +1,332 @@
+
+Additionally a validator should also be allowed to change values in a structural
+way to make coercing of values possible. Examples are casting a string to an
+integer or normalizing a unicode string.
+
+
+How it is currently done
+------------------------
+
+Currently there are many ways to specify field validation in Django. The easiest
+is to pass validator functions to the field via the ``validators`` argument:
+
+.. code-block:: python
+
+    title = CharField(validators=[
+        validate_illegal_characters,
+        validate_banned_words,
+    ])
+
+
+Where the validation functions just take a value and raise a ``ValidationError``
+if applicable.
+
+Custom fields can also override ``clean()``. This method can also change the
+value that is validated.
+
+.. code-block:: python
+
+    class TitleField(CharField):
+        def clean(self, value):
+            value = super().clean(value)
+            if not value.startswith('Title'):
+                # all titles must start with "Title"
+                value = 'Title ' + value
+            if len(value.split()) > 5:
+                raise ValidationError(
+                    'title must not contain more than 5 words'
+                )
+            return value
+
+
+It is also possible to define field validators on a form by adding
+``clean_<field_name>()`` methods to it:
+
+.. code-block:: python
+
+    class BlogPostForm(Form):
+        title = CharField()
+        content = CharField()
+
+        def clean_title(self):
+            title = self.cleaned_data['title']
+            if 'buzzword' in title.lower():
+                raise ValidationError('invalid word')
+            return title
+
+
+Just like a field's ``clean()`` method this method can also change the value.
+
+Cross field validation is made possible by overriding ``Form.clean()``:
+
+.. code-block:: python
+
+    class NumbersForm(Form):
+        first_number = IntegerField()
+        second_number = IntegerField()
+
+        def clean(self):
+            data = super().clean()
+            if data['first_number'] + data['second_number'] > 100:
+                raise ValidationError(
+                    'sum of numbers must be smaller than 100'
+                )
+            return data
+
+
+How it could be done
+--------------------
+
+There are several libraries that explicitly deal with validation in Django.
+
+
+Rendering
+=========
+
+Whenever the universal data structure or a subset thereof should be send or
+displayed to a user or an external service it must first pass the rendering
+step. It takes the universal data structure and converts it to something that
+can then be used by users (e.g. an HTML form) or services (e.g. a JSON object).
+
+How it is currently done
+------------------------
+
+Django uses the ``Widget`` class to render HTML form input elements. There is
+also the ``BoundField`` class that can be used in templates to customize how
+they are displayed. It is returned by ``Form.__getitem__()``:
```

Comment:
Also worth mentioning that rendering rules (such as help text, verbose name) can be passed down from many places to the widget.